### PR TITLE
Add persistence acknowledgement support for shipper

### LINKED ISSUE #32329: Make the shipper output to acknowledge batches based on `PersistedIndex` not `AcceptedCount`
**Describe the enhancement:**

Currently, [the shipper is acknowledging](https://github.com/elastic/beats/blob/06bf5dafec2de3ef0d3d858bd0f84cd7864b441d/libbeat/outputs/shipper/shipper.go#L175-L178) the event batch based on [`AcceptedCount`](https://github.com/elastic/elastic-agent-shipper-client/blob/de07abfef00ae4e1cd5012a1ff216d3c511fe3c5/api/messages/publish.proto#L68-L71) coming from the shipper as a part of the `PublishReply`.

This is unsafe because `AcceptedCount` does not mean that the actual final output (e.g. Elasticsearch) acknowledged these events from the Shipper's queue, it just means that the shipper put it in its own queue for processing. So, the data can be lost if the shipper process dies.

To fix this we need to use [`PersistedIndex`](https://github.com/elastic/elastic-agent-shipper-client/blob/de07abfef00ae4e1cd5012a1ff216d3c511fe3c5/api/messages/publish.proto#L80-L82) from `PublishReply` or a separate [`PersistedIndex` endpoint](https://github.com/elastic/elastic-agent-shipper-client/blob/de07abfef00ae4e1cd5012a1ff216d3c511fe3c5/api/shipper.proto#L29-L30) to acknowledge the events.

### PR DESCRIPTION
**Pull request by rdner**
## What does this PR do?

Now the shipper output is able to receive acknowledgement of the
persisted events from the shipper server.

Also, added the UUID verification in case the client is reconnected to
a different server.

## Why is it important?

To avoid data loss.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
~~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Related issues

- Closes #32329

### CODE REVIEW

File: libbeat/outputs/shipper/shipper.go [hunk 1]
```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
-	// the server is supposed to retain the order of the initial events in the response
-	// judging by the size of the result list we can determine what part of the initial
-	// list was accepted and we can send the rest of the list for a retry
-	retries := nonDroppedEvents[resp.AcceptedCount:]
+	if publishReply.PersistedIndex < publishReply.AcceptedIndex {
+		ackClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+			PollingInterval: durationpb.New(c.config.AckPollingInterval),
+		})
+		if err != nil {
+			return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
```
Comment (rdner) [inline libbeat/outputs/shipper/shipper.go:200]: Not sure if it's the right behaviour, this can cause data duplication.

File: libbeat/outputs/shipper/shipper.go [hunk 2]
```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
+		}
+
+		for {
+			indexReply, err := ackClient.Recv()
+			if err != nil {
+				return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
```
Comment (rdner) [inline libbeat/outputs/shipper/shipper.go:206]: Same here, not sure, another possible data duplication.

**Comment by rdner** [comment]
E2E test failure is not related to the change:

> [2022-08-16T17:23:54.410Z] time="2022-08-16T17:23:52Z" level=fatal msg="could not add package to policy; API status code = 400; response body = {\"statusCode\":400,\"error\":\"Bad Request\",\"message\":\"Package policy is invalid: inputs.logfile.streams.system.auth.vars.preserve_original_event: Preserve original event is required\\ninputs.logfile.streams.system.auth.vars.tags: Tags is required\"}"

File: libbeat/outputs/shipper/shipper.go [hunk 3]
```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
 		)
 	}
 
-	// the server is supposed to retain the order of the initial events in the response
-	// judging by the size of the result list we can determine what part of the initial
-	// list was accepted and we can send the rest of the list for a retry
-	retries := nonDroppedEvents[resp.AcceptedCount:]
+	if publishReply.PersistedIndex < publishReply.AcceptedIndex {
+		ackClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+			PollingInterval: durationpb.New(c.config.AckPollingInterval),
```
Comment (cmacknz) [inline libbeat/outputs/shipper/shipper.go:197]: We need to be careful here as the polling interval will effectively set a limit on the maximum throughput per worker. 

Per https://github.com/elastic/elastic-agent-shipper/pull/95#discussion_r947897465 we will probably always need to make this RPC call, and it is likely that we will need at least one polling interval to elapse to see the update we want. The minimal delay between receiving the `PublishResponse` and making this RPC may not be enough for the value of PersistentIndex to update. In the worst case we always incur at least one `AckPollingInterval` duration before being acknowledged.

If we keep the current `5 ms` polling interval then if we were to always incur that delay for every batch then we could publish at most 1000 ms / 5 ms = 200 batches / sec. 

With the default [max_bulk_size](https://www.elastic.co/guide/en/beats/filebeat/current/elasticsearch-output.html#_bulk_max_size) of 50 events this would limit the default throughput per worker to 200 * 50 = 10000 events/second. That doesn't feel unreasonable, but assuming my calculation is correct here we should document this type of calculation as the approximate throughput limit per shipper worker.

**Review by cmacknz — COMMENTED** [review]
Mostly looks good. I think we'll need a way to benchmark throughput between a beat and the shipper to really know how well this performs.

File: libbeat/outputs/shipper/shipper.go [hunk 4]
```diff
@@ -109,11 +111,28 @@ func (c *shipper) Connect() error {
...
+		PollingInterval: durationpb.New(0), // no need to subscribe, we need it only once
+	})
+	if err != nil {
+		return fmt.Errorf("failed to connect to the server: %w", err)
+	}
+	serverInfo, err := indexClient.Recv()
+	if err != nil {
+		return fmt.Errorf("failed to fetch server information: %w", err)
+	}
+	c.serverID = serverInfo.GetUuid()
```
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:128]: The first call to `PublishEvents` actually requires the ID as one of the request parameters and the server validates it against its own value.

File: libbeat/outputs/shipper/shipper.go [hunk 5]
```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
 	ctx, cancel := context.WithTimeout(ctx, c.timeout)
 	defer cancel()
-	resp, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
+	publishReply, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
+		Uuid:   c.serverID,
 		Events: convertedEvents,
 	})
 
-	if status.Code(err) != codes.OK || resp == nil {
+	if status.Code(err) != codes.OK || publishReply == nil {
```
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:180]: In theory, we can return `nil` from the gRPC endpoint by mistake, I just don't want the client to panic in this case.

File: libbeat/outputs/shipper/shipper.go [hunk 6]
```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
-			resp.AcceptedCount,
+			publishReply.AcceptedCount,
 		)
 	}
 
-	// the server is supposed to retain the order of the initial events in the response
-	// judging by the size of the result list we can determine what part of the initial
-	// list was accepted and we can send the rest of the list for a retry
-	retries := nonDroppedEvents[resp.AcceptedCount:]
+	if publishReply.PersistedIndex < publishReply.AcceptedIndex {
```
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:195]: Do you mean I should remove this condition? I mean even taking the comment into consideration it still works right.

File: libbeat/outputs/shipper/shipper.go (hunk 3 above)
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:197]: I'll add this to the config documentation.

File: libbeat/outputs/shipper/shipper.go [hunk 7]
```diff
@@ -153,37 +172,67 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
+		for {
+			indexReply, err := ackClient.Recv()
+			if err != nil {
+				return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+			}
+
+			if indexReply.GetUuid() != c.serverID {
+				batch.Cancelled()
+				st.Cancelled(len(events))
+				err := fmt.Errorf("acknowledgement failed due to a connection to a different server %s, expected %s", indexReply.Uuid, c.serverID)
```
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:212]: If I understood the code correctly, it's handled by this and the error is logged https://github.com/elastic/beats/blob/91775be8b2592ec6995ee339b17e39fda65d7f5d/libbeat/publisher/pipeline/client_worker.go#L172-L179

File: libbeat/outputs/shipper/shipper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go]: Removed in https://github.com/elastic/beats/pull/32708/commits/8a2453c89eb61c7129f791eb06ee50fb4a81a3a5

File: libbeat/outputs/shipper/shipper.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go]: Removed in https://github.com/elastic/beats/pull/32708/commits/509720590323e9bfc4c405c56a9489738da138b3

File: libbeat/outputs/shipper/shipper.go (hunk 3 above)
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go]: I added the calculations in https://github.com/elastic/beats/pull/32708/commits/7940a0d519b04595d0a230af35cb83ec01e4ea90

File: libbeat/outputs/shipper/shipper_test.go [hunk 8]
```diff
@@ -407,6 +385,45 @@ func TestPublish(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, expSignals, batch.Signals)
 	})
+
+	t.Run("cancel the batch when a differnet server responds", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper_test.go]: Fixed as a part of https://github.com/elastic/beats/pull/32708/commits/509720590323e9bfc4c405c56a9489738da138b3#diff-0741f72f20cb6abd66e45574ebf9f830e49ab64f857d921c375f9cdf83e8bb9bR389

File: libbeat/outputs/shipper/shipper.go (hunk 3 above)
Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: I'm not sure this is the right way to approach it. There's no reason to wait for full acknowledgment after every batch. Doing that would mean that no matter how big we make the shipper queue, only 1 batch worth of it would ever be filled, because the input won't advance until everything in it is sent.

But accumulating pending events is what the shipper queue is for -- if a batch has been published, we can immediately move on to the next batch and publish that, even if it isn't persisted yet. In that case the client would need to keep a list of outstanding batches, and if the persisted index is being removed from the publish response then it also needs a separate goroutine listening to persisted index changes and acknowledging batches as it goes.

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: (When I said in the meeting that this input needs to do things that most inputs won't, it was this: most clients can just call publish continuously and check the persisted index passively if at all. Beats is a special case because it _needs_ to advance the persisted index eventually or it gets stuck, because its own queue fills up. For inputs like filestream or kafka where the data source is relatively stable, this is redundant -- there's no particular maximum distance between the accepted and persisted indices, because there's no need for the client to keep a copy of the data. It's only because this adapter needs to handle arbitrary data from any beats queue that this case is more finicky.)

File: libbeat/outputs/shipper/shipper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (faec) [reply libbeat/outputs/shipper/shipper.go:128]: I thought the ID was optional? The intention was that equality was only enforced if the client included the ID, so that clients connecting for the first time (or clients that don't care about ACKs) could omit it

File: libbeat/outputs/shipper/shipper.go (hunk 3 above)
Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go]: > I'm not sure this is the right way to approach it. There's no reason to wait for full acknowledgment after every batch. Doing that would mean that no matter how big we make the shipper queue, only 1 batch worth of it would ever be filled, because the input won't advance until everything in it is sent.
>
> But accumulating pending events is what the shipper queue is for -- if a batch has been published, we can immediately move on to the next batch and publish that, even if it isn't persisted yet

This is an excellent point. We can and should take the position that the shipper client's job here is to offload data to the shipper as quickly as it can to minimize or eliminate the need to buffer anything in the Beat itself.

I'm not familiar enough with the rest of the beats pipeline yet to know how hard it would be to deal with asynchronously observing the persisted index and acknowledging batches. I can definitely see why it was done this way first as its the most natural way to satisfy the `Publish` interface.

File: libbeat/outputs/shipper/shipper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go:128]: It is optional: https://github.com/elastic/elastic-agent-shipper/blob/34571c41e892f8498d817a3b2784cc372b0985a2/server/server.go#L108

We can omit it in the first call as Fae described, we don't care if the shipper restarts between this point and the first call to `PublishEvents` (at least not from the perspective of acknowledgement tracking).

File: libbeat/outputs/shipper/shipper.go (hunk 3 above)
Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: I think it's a matter of returning immediately from the publish call (without updating the persisted index or acking / cancelling the batch) -- then save the outstanding batches in `shipper` and run a `PersistedIndex` listener in parallel, and when an update comes in _then_ call `batch.ACK()` on the batches that are covered. That way the beats pipeline will keep sending us batches after we finish the publish RPC.

However, I don't see a way around buffering things in the beat -- this approach will duplicate every pending event in memory (once in the beats queue and once in the shipper queue) because the beats queue can't safely remove an event until it's persisted, and that doesn't happen until the shipper is also done with it.

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: Actually, I do see a way around buffering things in the beat, but it'll take some fiddling with the pipeline: we could give this specific output a special queue configuration so its events are freed when they are published, but the ACKs aren't sent to the input producer until they're persisted. It will take some plumbing but probably worth it to avoid a 2x memory use.

But that is independent of this PR -- here, the ideal behavior is still to send things as fast as you can and acknowledge outstanding batches when their index is persisted.

Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go]: > Actually, I do see a way around buffering things in the beat, but it'll take some fiddling with the pipeline: we could give this specific output a special queue configuration so its events are freed when they are published, but the ACKs aren't sent to the input producer until they're persisted. It will take some plumbing but probably worth it to avoid a 2x memory use.

Excellent, can you create an issue for us to do this as a follow up? I definitely want to avoid a 2x increase in memory usage with the shipper.

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: https://github.com/elastic/elastic-agent-shipper/issues/97

Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go]: Thanks! I think my preference right now is to keep this implementation as is and come back to it with the changes to https://github.com/elastic/elastic-agent-shipper/issues/97.

I think we can have an experimental shipper with this client implementation. Considering where we are in the release cycle I'd rather Denis focus on getting some integration tests written to validate the system and find any problems we don't know about yet.

Comment (rdner) [reply libbeat/outputs/shipper/shipper.go]: > I think it's a matter of returning immediately from the publish call (without updating the persisted index or acking / cancelling the batch) -- then save the outstanding batches in shipper and run a PersistedIndex listener in parallel, and when an update comes in then call batch.ACK() on the batches that are covered. That way the beats pipeline will keep sending us batches after we finish the publish RPC.

@faec this is exactly the approach I took first but it got really complicated very quick  and I remembered we said that the client must be as simple as possible for other developers to implement. So, that's why I ended up with this blocking implementation instead.

Doing this async would make us to do at least the following:

1. Define this new type
```go
 type pendingBatch struct {                                                                                                                                                                                                                                                   
   batch            publisher.Batch
   serverID         string
   acceptedCount    uint32
   acceptedIndex    uint64
   nonDroppedEvents []publisher.Event
 }
```
2. Once a batch is accepted on the server we send a `pendingBatch` to `chan pendingBatch` in the output's `Publish` function
3. Listen to changes of `PersistedIndex` and `UUID` using the `PersistedIndex` endpoint in a background go routine
4. Listen to a buffered `chan pendingBatch` in a background go routine
5. On each `pendingBatch` from `chan pendingBatch` we block until a signal from 3. (e.g. using `sync.Cond`) when `PersistedIndex` >= `pendingBatch.AcceptedIndex` with all the synchronisation difficulties involved

Also, there are edge cases to handle:
1. When not entire batch was accepted and we must retry the rest
2. When multiple pending batches should be acknowledged from a single `PersistedIndex` update
3. When the server UUID changes and we need to cancel all pending batches sent to this server

and perhaps many more that I didn't hit yet.

So, if our intention to keep the client simple it has to be synchronous or we stop pursuing this goal and make it complex but high-performant.

File: libbeat/outputs/shipper/shipper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go:128]: I think this can be removed to simplify the connection logic. The first `PublishEvents` RPC can have an empty server ID and everything will work correctly. We need to save the shipper ID return from that first call though.

File: libbeat/outputs/shipper/shipper.go (hunk 3 above)
Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: Well, simplicity is nice but I think limiting the entire shipper queue to at most one batch is enough of a constraint that it's worth tracking some metadata.

However, I think it can be at least somewhat simpler than your suggestion. I would do something like:

(1) add a list of pending batches to the shipper struct

```go
type shipper struct {
	...
	pending []struct {
		batch publisher.Batch
		index uint64
	}
	pendingLock sync.Mutex
}
```

(2) modify `shipper.Publish` so it sends _all_ events in the batch (making multiple upstream publish calls if necessary), and saves the batch as well as the _final_ accepted index in `shipper.pending`. [This is more in keeping with the intent of the API anyway: currently the shipper only blocks until at least one event is sent, but that's just because the queue is full, not because some of the events really "failed." So marking these events as failed and needing retry in the beats output doesn't fit the usual semantics -- we don't want to be potentially dropping lots of events over "repeated failures" when all that's really happening is back pressure.]

(3) Add a continuous listener to `PersistedIndex`, and when a new value comes in do something like:

```go
s.pendingLock.Lock()
batchCount := 0
while batchCount < len(s.pendingLock) && s.pending[batchCount].index < newPersistedIndex {
	s.pending[batchCount].ACK()
	batchCount++
}
s.pending = s.pending[batchCount:]
s.pendingLock.Unlock()
```

We can jump in a zoom to talk through this if you want, I know the pipeline logic is a little arcane and this is getting long.

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: Actually, one more small change that would help with freeing up memory: instead of saving the original `Batch` structure, just save the `ACK` callback. Fingers crossed, but that might be enough to keep pending events from hanging around in memory once we make the appropriate queue changes.

Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go]: > Well, simplicity is nice but I think limiting the entire shipper queue to at most one batch is enough of a constraint that it's worth tracking some metadata.

I think I agree that having a simple reference client is nice, but we also need the Beats client to be performant to ensure users experience use of the shipper as a strict improvement over what we had before. Tiago has started integrating the shipper in the V2 inputs repository and since that is a from scratch system it may be better suited as the reference client.

I think we need both follow @faec's suggestion above to make the listening async and also  make the change described in https://github.com/elastic/elastic-agent-shipper/issues/97 to ensure we do not double buffer events (once in beats, once in the shipper).

I don't know that we need to do both of these things in this PR however, it is fine as a starting point. Certainly the queue changes can be separate.

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: The queue changes are definitely separate. Making this code async could be an in-place revision or a second pass, but should definitely happen before we give this output to customers. (But since we're pre-beta I would be willing to approve this one if there's a followup issue.)

Comment (rdner) [reply libbeat/outputs/shipper/shipper.go]: I'll make it async in a follow up PR, @faec do you think we can merge the current state then? I'm still waiting for an approval :)

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: Well, I wanted an explicit issue before approving, so I wrote one 😜 https://github.com/elastic/elastic-agent-shipper/issues/98 I will do one last pass of the rest of this PR

File: libbeat/outputs/shipper/shipper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:128]: I'm not sure about this. `Connect` can be called multiple times by the beats pipeline:

https://github.com/elastic/beats/blob/91775be8b2592ec6995ee339b17e39fda65d7f5d/libbeat/publisher/pipeline/client_worker.go#L133-L159

This means that every time we have a connection error, we would re-use a previous serverID and the next batch publication would fail with the validation error which would cause an infinite loop if I removed this logic from `Connect`. I think it's more robust this way.

[the remark this answers is not in the document]
Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go:128]: Ah I didn't realize that just reading the code here in isolation, can you add a comment clarifying this?

[the remark this answers is not in the document]
Comment (faec) [reply libbeat/outputs/shipper/shipper.go:128]: I'm not sure I follow the conflict here -- can't `Connect` just set `c.serverID` to an empty string and skip the whole `PersistedIndex` call? That will mean if there's a connection error, the server id will be reset, which is all we really need to keep the id from being reused, and `Publish` calls need to check the returned id anyway.

File: libbeat/outputs/shipper/shipper.go [hunk 9]
```diff
@@ -153,37 +172,65 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
...
+		indexReply, err := ackClient.Recv()
+		if err != nil {
+			return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+		}
+
+		if indexReply.GetUuid() != c.serverID {
+			batch.Cancelled()
+			st.Cancelled(len(events))
+			err := fmt.Errorf("acknowledgement failed due to a connection to a different server %s, expected %s", indexReply.Uuid, c.serverID)
+			c.serverID = indexReply.GetUuid()
```
Comment (faec) [inline libbeat/outputs/shipper/shipper.go:212]: Following on the other comment about `serverID` initialization, this `serverID` update has no effect since it will be immediately overwritten by a reconnect

Comment (faec) [reply libbeat/outputs/shipper/shipper.go:212]: How about moving this `if` block out of the loop with the first `Recv` call so we only do this once? This is all one RPC call so the server ID can't change between iterations.

Comment (rdner) [reply libbeat/outputs/shipper/shipper.go:212]: @faec well, as you said this line has no effect because of the pipeline logic which is supposed to call `Connect` again. Why moving this to somewhere else if we can remove?

Also, none of this logic and edge cases are final. We'll find out much more after the integration tests are done.

Comment (faec) [reply libbeat/outputs/shipper/shipper.go]: For removal I was talking about `serverID` but the moving was the `if` block, `if indexReply.GetUuid() != c.serverID`, since I assume we do want to check server id once but there's no need to check it multiple times for a stable connection -- if the first read is before the infinite loop, then the rest of the loop can do it the simple way.

Comment (rdner) [reply libbeat/outputs/shipper/shipper.go]: @faec I moved the check in https://github.com/elastic/beats/pull/32708/commits/f6795c3ff42c906d7ed763ff73a61b43f8ff5b15

I also realised that `CloseSend` is not the right way to close the stream, the gRPC doc requests to cancel the context, so I did that too.

File: libbeat/outputs/shipper/shipper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmacknz) [reply libbeat/outputs/shipper/shipper.go:128]: @rdner I think this is the only unresolved thread, was it addressed with the most recent changes?

I approved as I don't think this is blocking given that we have some follow up changes coming.
